### PR TITLE
Stats: Move the Post Stats API calls to the data layer

### PR DESCRIPTION
**Pull request by youknowriad**
This is just a small PR to get my hands on the [data layer](https://github.com/Automattic/wp-calypso/tree/5e898c923d1fa01cc955bdfa68461e21ec8a540f/client/state/data-layer)

So In this PR, I'm moving the API calls for the Post Stats to the data layer. I have some questions left:
 - The former action creator used to trigger a "request" action, this was used to update the "isRequesting" flag. If I understand correctly, in the data layer, the "request" action does not make its way into the Redux Store. Thus, I used a `POST_STATS_REQUEST_START` instead. Is this the correct way to deal with this.
 - I asked myself if I should move the `receivePostStat` action creator and I've decided to keep it outside the data layer. My rule was, if it's just a synchronous action creator, keep it outside the data layer. Is this the correct way of thinking?
 - I admit that for now, it feels just like moving things from a file to another and I would like to add some advanced behaviour here, like "refreshing". Currently, the refreshing for PostStats is handled on the `QueryPostStats` Component. as AFAIK, one of the goals of the data layer is to handle this, Is there any example (or ideas) for this?

**Testing Instructions**

- Open the site stat insights page 
- The "Last Post Summary" should show the views count correctly

### CODE REVIEW

**Comment by dmsnell** [comment]
> The former action creator used to trigger a "request" action, this was used to update the "isRequesting" flag. If I understand correctly, in the data layer, the "request" action does not make its way into the Redux Store. Thus, I used a POST_STATS_REQUEST_START instead. Is this the correct way to deal with this.

Whether or not the `isRequesting` flag goes into the Redux store is now more of a question of whether or not the app needs to render something to indicate that content is loading, though ideally that is already captured in some other form, such as a `null` value in state and a duplicate flag isn't necessary.

There's not particularly a _correct way_ to do any of this yet as we're just starting to figure it out. I would suggest that we don't need a specific "start" action. The components should not be instructing the data layer to do anything, only announcing that they have a need.

There could be other indications to put into the Redux store to suggest that data is loading. At this point I think we have the ability to start thinking about what those notions are or should be instead of focussing on "an HTTP request is in flight."

**Comment by dmsnell** [comment]
> I asked myself if I should move the receivePostStat action creator and I've decided to keep it outside the data layer. My rule was, if it's just a synchronous action creator, keep it outside the data layer. Is this the correct way of thinking?

That's basically right since Redux is a _synchronous_ data system. `receivePostStat` however does seem to imply that it's getting a response from somewhere. Maybe for now it's best to leave it as-is as I suspect that this has been here a while, but ideally we would map the API response into a Calypso-centric data format and then use some action like `addPostStat` or `updatePostStat` to show that we're injecting actual data, not blindly handing off API data. Mapping brings a host of benefits, one of which I recently discovered is the triviality of injecting UI state into the app. If we know the shape of the data that the state expects then we can `dispatch( { type: actionTypes.POST_STAT_ADD, siteId, postId, stats } )` all from the command line, having that code saved in a buffer somewhere.

**Comment by dmsnell** [comment]
> I admit that for now, it feels just like moving things from a file to another and I would like to add some advanced behaviour here, like "refreshing". Currently, the refreshing for PostStats is handled on the QueryPostStats Component. as AFAIK, one of the goals of the data layer is to handle this, Is there any example (or ideas) for this?

This feeling may be slow to go away as we figure out how to use this properly. A query component whose underlying data is being handled by the data middleware should not need to do anything but announce that it needs its data and then we can do just what you suggested. Maybe the right behavior to add into the middleware is like this…

 - When we receive a request for data, start polling that data
 - When we receive our fetch responses, check when the last time was that we received a request for data
 - If no request has come in during the past N minutes/seconds, assume that the UI no longer needs it to be fresh and stop polling.

**Comment by dmsnell** [comment]
@youknowriad I want to thank you and let you know I appreciate you pushing into the data layer. this is all still new and we're figuring out the means and norms and edges together. thanks for asking those questions and please be liberal with further questions and discussions.

File: client/state/data-layer/wpcom/sites/stats/post/index.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	} ).catch( error => {
+		dispatch( {
+			type: POST_STATS_REQUEST_FAILURE,
+			stat,
+			siteId,
+			postId,
+			error
+		} );
+	} );
+};
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/stats/post/index.js]: This is a good place to think about using the `BATCH` endpoint. I don't know how the post stat requests go out, but they are simple `GET` requests aren't they? In such a case we could queue these up in increments of say, 50ms, then send them all together in one.

```js
const queue = ( store, action ) => {
	fetchQueue.push( action );
	if ( fetchQueue.olderThan( 50 ) ) {
		requests = fetchQueue.flush();
		wpcom
			.batch( requests.map( toRequestUrl ) )
			.then( responses => responses.forEach( response.dispatch( receivePostStat( response ) ) ) );
	}
}

export default {
	[ POST_STATS_REQUEST ]: [ queue ],
}
```

please don't use this code. it's only meant to illustrate the idea

Comment (youknowriad) [inline client/state/data-layer/wpcom/sites/stats/post/index.js]: Great idea! I pushed a first attempt here 4a5eb36

It's working correctly but the `batch` endpoint could be improved: For now, when an URL returns an error, the status_code is included in the response, but when it's successful, it's not. I think it should behave consistently and always include an "envelope" to the response.

Comment (youknowriad) [inline client/state/data-layer/wpcom/sites/stats/post/index.js]: The test is failing, because I could not manage to mock the batch URI using `nock`. I was thinking about an URL encoding issue, but no matter what I try, I could not make it work for now.

Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/stats/post/index.js]: Yeah I'm thinking this may be a good idea for another PR to enhance the behavior where we can isolate the changes to introducing the batching. I like what's happening but I don't think we're using this at all in Calypso yet so there are definitely things to figure out, such as what you found out with testing.

Comment (youknowriad) [inline client/state/data-layer/wpcom/sites/stats/post/index.js]: I moved it to its own PR https://github.com/Automattic/wp-calypso/pull/10270 :)

**Comment by youknowriad** [comment]
I'm closing this for now. This action creator has changed a lot and it's probably better to restart from scratch for this Endpoint. (We may consider merging this subtree with stats lists aswell)
